x-pack/plugins/security_solution/public/timelines/store/middlewares/timeline_favorite.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { get } from 'lodash/fp';
+import type { Action, Middleware } from 'redux';
+import type { CoreStart } from '@kbn/core/public';
+
+import type { State } from '../../../common/store/types';
+import {
+  endTimelineSaving,
+  updateIsFavorite,
+  updateTimeline,
+  startTimelineSaving,
+  showCallOutUnauthorizedMsg,
+} from '../actions';
+import type { ResponseFavoriteTimeline } from '../../../../common/api/timeline';
+import { TimelineType } from '../../../../common/api/timeline';
+import { persistFavorite } from '../../containers/api';
+import { selectTimelineById } from '../selectors';
+import * as i18n from '../../pages/translations';
+import { refreshTimelines } from './helpers';
+
+type FavoriteTimelineAction = ReturnType<typeof updateIsFavorite>;
+
+function isFavoriteTimelineAction(action: Action): action is FavoriteTimelineAction {
+  return action.type === updateIsFavorite.type;
+}
+
+export const favoriteTimelineMiddleware: (kibana: CoreStart) => Middleware<{}, State> =
+  (kibana: CoreStart) => (store) => (next) => async (action: Action) => {
+    // perform the action
+    const ret = next(action);
+
+    if (isFavoriteTimelineAction(action)) {
+      const { id } = action.payload;
+      const timeline = selectTimelineById(store.getState(), id);
+
+      store.dispatch(startTimelineSaving({ id }));
+
+      try {
+        const result = await persistFavorite({
+          timelineId: timeline.id,
+          templateTimelineId: timeline.templateTimelineId,
+          templateTimelineVersion: timeline.templateTimelineVersion,
+          timelineType: timeline.timelineType ?? TimelineType.default,
+        });
+
+        const response: ResponseFavoriteTimeline = get('data.persistFavorite', result);
+
+        if (response.code === 403) {
+          store.dispatch(showCallOutUnauthorizedMsg());
+        }
+
+        refreshTimelines(store.getState());
+
+        store.dispatch(
+          updateTimeline({
+            id,
+            timeline: {
+              ...timeline,
+              isFavorite: response.favorite != null && response.favorite.length > 0,
+              savedObjectId: response.savedObjectId || null,
+              version: response.version || null,
+            },
+          })
+        );
+      } catch (error) {
+        kibana.notifications.toasts.addDanger({
```

Comment:
Do you think it would be better to put the notifications in the UI components instead? Then we wouldn't need to pass `kibana$` at all. This may be a better idea though once we decide on how we want to refactor our queries...

Reply:
It depends. I like that this side-effect doesn't live in five different components. Having this non-consequential side-effect in a middleware made the most sense to me. Passing `kibana` isn't that much of a hassle, imo. Wdyt?